### ISSUE TITLE
Interpolation not effective in some cases

### ISSUE THREAD

**Issue by myllynen** [performancecopilot/pcp#433]
As discussed in https://groups.io/g/pcp/topic/6111366, when creating an archive with pmlogger(1) the prologue is logged ~0.02 seconds before the actual data and we see for example with pmdumptext(1) shows results being 10 minutes off and misses some records in the end. I'm providing a copypaste and a test archive to demonstrate this issue below.

```
$ pmdumplog -z -a test                                          
Note: timezone set to local timezone of host "test" from archive

Log Label (Log Format Version 2)
Performance metrics from host test
    commencing Fri Oct 20 10:36:00.416347 2017
    ending     Fri Oct 20 11:26:00.436134 2017
Archive timezone: EEST-3
PID for pmlogger: 6479

Descriptions for Metrics in the Log ...
PMID: 60.1.2 (mem.util.free)
    Data Type: 64-bit unsigned int  InDom: PM_INDOM_NULL 0xffffffff
    Semantics: instant  Units: Kbyte
PMID: 2.3.2 (pmcd.pmlogger.archive)
    Data Type: string  InDom: 2.1 0x800001
    Semantics: discrete  Units: none
PMID: 2.3.0 (pmcd.pmlogger.port)
    Data Type: 32-bit unsigned int  InDom: 2.1 0x800001
    Semantics: discrete  Units: none
PMID: 2.3.3 (pmcd.pmlogger.host)
    Data Type: string  InDom: 2.1 0x800001
    Semantics: discrete  Units: none

Instance Domains in the Log ...
InDom: 2.1
10:36:00.416347 1 instances
   6479 or "6479"

Temporal Index
		Log Vol    end(meta)     end(log)
10:36:00.416347	      0          132          132
11:26:00.436134	      0          403          528

[136 bytes]
10:36:00.416347 3 metrics
    2.3.3 (pmcd.pmlogger.host): inst [6479 or "6479"] value "localhost"
    2.3.0 (pmcd.pmlogger.port): inst [6479 or "6479"] value 4330
    2.3.2 (pmcd.pmlogger.archive): inst [6479 or "6479"] value "/home/testuser/pcptest/logger/test"

[52 bytes]
10:36:00.435851 1 metric
    60.1.2 (mem.util.free): value 6925664

[52 bytes]
10:46:00.436619 1 metric
    60.1.2 (mem.util.free): value 3764672

[52 bytes]
10:56:00.436598 1 metric
    60.1.2 (mem.util.free): value 3374424

[52 bytes]
11:06:00.436446 1 metric
    60.1.2 (mem.util.free): value 5832004

[52 bytes]
11:16:00.436239 1 metric
    60.1.2 (mem.util.free): value 5574748

[52 bytes]
11:26:00.436134 1 metric
    60.1.2 (mem.util.free): value 5685308
$ pmdumptext -z -a test -t 10min mem.util.free                  
Fri Oct 20 10:36:00	?
Fri Oct 20 10:46:00	6925664.000
Fri Oct 20 10:56:00	3764672.000
Fri Oct 20 11:06:00	3374424.000
Fri Oct 20 11:16:00	5832004.000
Fri Oct 20 11:26:00	5574748.000
$ 
```

[test-archive.tar.gz](https://github.com/performancecopilot/pcp/files/1681743/test-archive.tar.gz)

Let's use this issue to document expectations and track possible fixes around this. Thanks.

**Comment by kmcdonell** [comment]
@myllynen I'm experimenting with a change to pmlogger that will defer writing the label records and the preamble (which is a synthetic pmResult) until the first real pmResult is seen, then using the timestamp from the first pmResult in the label records and the preamble (so we'll get two pmResults at the start with the same timestamp).  This certainly fixes your pmdumptext example, but I need to be sure there are no other unexpected side-effects.

**Comment by myllynen** [comment]
@kmcdonell I have missed your comment here earlier. Did you have a chance to test this further, did you come across any side-effects? Thanks.

**Comment by kmcdonell** [comment]
G'day @myllynen thanks for prodding me on this issue ... I've unpacked your test archive and I really don't think there is an issue here (although there is an element of "not as expected" without an explanation), and I would suggest this issue be closed.

The "experiment" I referred to above did not get implemented, as it turned into a rat hole convoluting pmlogger even more and triggers all manner of unnatural acts because time is no longer monotonic increasing if two pmResults have the same timestamp.

To explain why I don't think there is an "issue" here, consider this timeline
```
---+-----+--------------+----- ... -----+------> time
   t(0)  t(1)           t(2)            t(N)
```
where the first pmResult in the archive is at time t(0) and the last pmResult in the archive is at time t(N).  For your test archive t(0) = 10:36:00.416347 and t(N) = t(6) = 11:26:00.436134.  Now the label record correctly shows the period covered by the archive as
```
    commencing Fri Oct 20 10:36:00.416347 2017
    ending     Fri Oct 20 11:26:00.436134 2017
```

Now consider some arbitrary metric A (mem.util.free in your case) and let's assume it has the semantics of "discrete" as this is the simplest case:

1. The metric A may first appear at time t(j), for some 0 <= j <= N, and no matter what tool is used there is NO value for A for time t in the range t <= t(j) ... so the ? from pmdumptext for the first reporting sample is correct and expected.
2. The presence of the "preamble" pmResult makes 1. universally true for any reporting tool that starts at the "commencing" time of the archive and is looking for metrics other than the pmcd.* ones in the preamble pmResult ... users could use a hand crafted -O or -S to get past this.
3. Since the pmlogger config may include many groups of disjoint metrics with different logging intervals, the first observation of metric A may not be in the pmResult at t(1), or (t2), or ....  so in general we don't need 2. to see 1. And there is no safe way to start anywhere but t(0), especially if the tool is asking for not just metric A but a handful of different metrics each of which could be first seen at a different t(j)
4. If you start at t(0), and report at regular intervals of K (secs, hrs, whatever), then after some S samples, we're at t(0) + S*K and the last pmResult in the archive is at t(N), such that t(0) + S*K <= t(N) < t(0) + (S+1)*K and we cannot report a value for sample S+1 .  For your example K=10min, S=5 and the last reported time is 10:36:00.416347 + 50min = 11:26:00.416347; the next time would be 11:36:00.416347 which is past the end of the archive, so nothing to report.  I don't see any "missing records at the end here" .
5. In general if you log A at a regular interval for some X samples, then replaying the archive with a sampling interval equal to the logging interval will result in values for fewer than X samples being reported, it may be X-1 as in your example above, but could be X-2 or even X-3 (the situation is worse for metrics with counter or instantaneous semantics).
6. If A has an instance domain then the arguments above apply to **every** instance of A.
7. If A has a dynamic instance domain (like the proc.* metrics), the problem just gets worse because instances appear and disappear to times we have no control over.

I think these are all "expected" boundary conditions at the beginning and end of an archive, and for most practical purposes would never be noticed.

I'll wait on your response before changing the status of this issue.

**Comment by myllynen** [comment]
Thanks for the detailed explanation. But I'm afraid this now comes back to what we discussed earlier in the above mentioned thread at https://groups.io/g/pcp/topic/6111366, see especially the last few messages there.

Also, the pmSetMode(3) manual page states:

      If  the mode is PM_MODE_INTERP then, in the case of pmFetch(3), the un‐
       derlying code will use an interpolation scheme to compute the values of
       the  metrics from the values recorded for times in the proximity of the
       time origin.  A mode of PM_MODE_INTERP may only be used with an archive
       context.

To me as a user it seems that there is not much interpolation happening given that the value which is a fraction of a second off from the reporting time is not considered at all while the value off by 10 minutes from the reporting time is being reported as-is. Or perhaps I'm misunderstanding something here or this all could related to the case of logging and reporting interval being the same which somehow causes this behavior that goes against my expectations. Or perhaps I just don't understand what that computation and proximity really refer to in the manual page.

In any case, I think I've explained my understanding in the referenced thread and here. Although I might well still be mistaken or ignoring some crucial low-level details I wouldn't be surprised if at least some other users - e.g. after reading pmSetMode(3) - could have similar expectations.

I'll leave it up to your consideration is there anything to change in the code (thinking more of interp.c here rather than triggering the unnatural acts with pmlogger) or in the man page(s) or neither. Thanks again for looking into this.

**Comment by kmcdonell** [comment]
Ahh @myllynen at last I am seeing your point (it has only taken 7 years).  The man page wording is misleading, especially the word "proximity" ... interpolation is used for counters, but extrapolation is used for other metric semantics and the value is based on the closest (in time) earlier value and later value.

A more precise definition is given in https://pcp.io/papers/archive-interpolation.pdf, where for instantaneous metrics the algorithm is:

The value for a metric with INSTANTANEOUS semantics at an arbitrary time t is:
• If there is some xa(tl) for a largest tl ≤ t, then this is the prior bound.
• If there is some xa(tu) for a smallest tu ≥ t, then this is the next bound.
• If both the prior bound and the next bound exist, then xf(t) is computed by **extrapolation**, i.e. xf(t)
= xa(tl), otherwise no value exists for xf(t).

the choice of extrapolation was based on the notion that in the absence of any new information, the "best" guess for a value is the **last** value we've seen in the direction of time travel (being able to reply actives backwards in time just complicates the whole implementation by a factor of 2!).

Now with hindsight, I can see that we may have been better to replace the last bullet above by

• If both the prior bound and the next bound exist, then xf(t) is computed by **proximity**, i.e. then xf(t)
= (tl+tu)/2 <= tl ? xa(tl) : xa(tu), otherwise no value exists for xf(t).

which would be "pick the closest in time" and closer to **proximity**.

I've had a look at the code and this is feasible (a cookie cutter change is needed in lots of places), but making the change just for the semantics of mem.util.free, your test case above now looks like:
```
$ LD_LIBRARY_PATH=/home/kenj/src/pcp/src/libpcp/src pmdumptext -z -a test -t 10min mem.util.free
Fri Oct 20 10:36:00	?
Fri Oct 20 10:46:00	3855024128.000
Fri Oct 20 10:56:00	3455410176.000
Fri Oct 20 11:06:00	5971972096.000
Fri Oct 20 11:16:00	5708541952.000
Fri Oct 20 11:26:00	5821755392.000

Is that closer to what you might have been expecting?

If so, we (the PCP developers) need to huddle and decide if this is a change we want to make.

And the pmSetMode man page needs some re-wording, independent of the decision.

But I do think the effects here are more obvious because you're replaying the archive with sample times that are identical to (or a multiple of) the archive logging intervals, so the sampling times are falling close to the archive log times.

**Comment by myllynen** [comment]
Thanks, now things start to look more understandable even for me!

But just to clarify, in your example if the values are from the above test archive they look like have been multiplied by thousand at some point? If I divide the values by thousand then the results make sense to me and are what I would expect.

If the decision eventually is to make this change I wonder would it be possible at the same time to consider some sort of "weight" in the formula depending the distance of values from the current reporting time? Or would it open a can of worms which looks (relatively) easy on the surface but would soon get too tricky to be feasible?

Thanks for the investigation, please feel free to ping me if you want me to review the upcoming man page tweaks.

Out of curiosity, as I'm still finding myself wondering, how did the earlier mentioned experiment lead to convoluted code? My most likely very naïve idea would have been to simply add "static int prologue_written = 0;" in pmlogger.c and then when adding metrics to the archive check if the prologue needs to be added at the same go or not, one if (prologue_written) wouldn't cost much and this in theory would eliminate the small distance between the prologue and the first sample. But since I haven't really looked at the code and don't know the side-effects of this perhaps the idea would work only in a dream world :-)

Thanks.

**Comment by kmcdonell** [comment]
@myllynen we're getting closer 😁 

Values are because pmdumptext reports in canonical units of bytes ... pmval reports in archive units of Kbytes.

Weight is not going to happen ... unlike counters, instantaneous metrics don't really allow any arithmetic reasoning about values between two consecutive observations.

The preamble issue is orthogonal to value sythesis, I'll take another look at it but my main concern is not deferred emitting of the preamble, but two pmResults with the same timestamp and having to rewrite the archive label records.

**Comment by myllynen** [comment]
@kmcdonell Thanks! So the only open question here would be the preamble issue.

Do you see it as something that could be still addressed before 7.0?

**Comment by kmcdonell** [comment]
@myllynen as of commit fe3d14777 I believe this is now done.
For example,  with a pmlogger configuration of
`log mandatory on 10sec hinv.ndisk`
in urk.config and pmlogger run as
`$ pmlogger -c urk.config -s 1 urk`
this archive dump
```
$ pmlogdump -lmt urk
Log Label (Log Format Version 3)
Performance metrics from host bozo.localdomain
    commencing Fri Aug 15 06:40:57.213471279 2025
    ending     Fri Aug 15 06:40:57.216918528 2025

Temporal Index
			Log Vol    end(meta)     end(log)
06:40:57.213471279	      0          808          808
06:40:57.216918528	      0         1604         1048

06:40:57.213471279 5 metrics
    2.3.3 (pmcd.pmlogger.host): inst [395109 or "395109"] value "bozo.localdomain"
    2.3.0 (pmcd.pmlogger.port): inst [395109 or "395109"] value 4331
    2.3.2 (pmcd.pmlogger.archive): inst [395109 or "395109"] value "/home/kenj/src/pcp/qa/tmp/urk"
    2.0.23 (pmcd.pid): value 82875
    2.0.24 (pmcd.seqnum): value 19

06:40:57.213471279 1 metric
    60.0.33 (hinv.ndisk): value 5

06:40:57.216918528 5 metrics
    2.3.3 (pmcd.pmlogger.host): inst [395109 or "395109"] value "bozo.localdomain"
    2.3.0 (pmcd.pmlogger.port): inst [395109 or "395109"] value 4331
    2.3.2 (pmcd.pmlogger.archive): inst [395109 or "395109"] value "/home/kenj/src/pcp/qa/tmp/urk"
    2.0.23 (pmcd.pid): value 82875
    2.0.24 (pmcd.seqnum): value 19
```
shows that the archive label record, the initial temporal index entry and the preamble pmResult all have the same timestamp as the first real pmResult for hinv.ndisk.

I'm closing this issue, but please re-open or (better) start another issue if you feel this is still NQR.
